### PR TITLE
Removed folderName from restore public APIs in favor of taking a whole URL containing including the folder name to later parse it.

### PR DESCRIPTION
**Pull request by vcolin7**
Based on [.NET's restore APIs](https://github.com/Azure/azure-sdk-for-net/pull/15885).

### CODE REVIEW

**Comment by vcolin7** [comment]
I'm not sure to what level we should validate the given `backupFolderUrl`. Assuming the user uses what is returned from running a backup operation, no validation should be necessary, but I wonder if it would be better to use either Java's `URI` or `URL` classes, since at least the use of the latter is [discouraged in the guidelines](https://azure.github.io/azure-sdk/java_design.html#java-api-guidance) at the public API level, which makes me think we might want to avoid it in the implementation too.

Any thoughts on this? @g2vinay

**Comment by christothes** [comment]
> I'm not sure to what level we should validate the given `backupFolderUrl`. Assuming the user uses what is returned from running a backup operation, no validation should be necessary, but I wonder if it would be better to use either Java's `URI` or `URL` classes, since at least the use of the latter is [discouraged in the guidelines](https://azure.github.io/azure-sdk/java_design.html#java-api-guidance) at the public API level, which makes me think we might want to avoid it in the implementation too.
> 
> Any thoughts on this? @g2vinay

I can't speak from the Java guidelines perspective, but to me it makes sense to be consistent with the return type of the backup operation. If there is a case to be made for making this a `URI` I'd think it should be changed in both places. 

That said, comparing to the storage sdk, they seem to use only `String` parameters to address blob endpoint urls. From that perspective, leaving it as `String` seems to be the most consistent choice.

File: sdk/keyvault/azure-security-keyvault-administration/src/main/java/com/azure/security/keyvault/administration/KeyVaultBackupAsyncClient.java [hunk 1]
```diff
@@ -340,20 +340,20 @@ private static LongRunningOperationStatus toLongRunningOperationStatus(String op
 
         return clientImpl.fullRestoreOperationWithResponseAsync(vaultUrl, restoreOperationParameters,
             context.addData(AZ_TRACING_NAMESPACE_KEY, KEYVAULT_TRACING_NAMESPACE_VALUE))
-            .doOnRequest(ignored -> logger.info("Restoring from location - {}\\{}", blobStorageUrl, folderName))
-            .doOnSuccess(response -> logger.info("Restored from location - {}\\{}", blobStorageUrl, folderName))
+            .doOnRequest(ignored -> logger.info("Restoring from location - {}{}", containerUrl, folderName))
+            .doOnSuccess(response -> logger.info("Restored from location - {}{}", containerUrl, folderName))
             .doOnError(error ->
-                logger.warning("Failed to restore at location - {}\\{}", blobStorageUrl, folderName, error))
+                logger.warning("Failed to restore from location - {}{}", containerUrl, folderName, error))
```
Comment (christothes) [inline sdk/keyvault/azure-security-keyvault-administration/src/main/java/com/azure/security/keyvault/administration/KeyVaultBackupAsyncClient.java:346]: nit: simplify to below?
```suggestion
                logger.warning("Failed to restore from location - {}", backupFolderUrl, error))
```

File: sdk/keyvault/azure-security-keyvault-administration/src/main/java/com/azure/security/keyvault/administration/KeyVaultBackupAsyncClient.java [hunk 2]
```diff
@@ -340,20 +340,20 @@ private static LongRunningOperationStatus toLongRunningOperationStatus(String op
 
         return clientImpl.fullRestoreOperationWithResponseAsync(vaultUrl, restoreOperationParameters,
             context.addData(AZ_TRACING_NAMESPACE_KEY, KEYVAULT_TRACING_NAMESPACE_VALUE))
-            .doOnRequest(ignored -> logger.info("Restoring from location - {}\\{}", blobStorageUrl, folderName))
-            .doOnSuccess(response -> logger.info("Restored from location - {}\\{}", blobStorageUrl, folderName))
+            .doOnRequest(ignored -> logger.info("Restoring from location - {}{}", containerUrl, folderName))
```
Comment (christothes) [inline sdk/keyvault/azure-security-keyvault-administration/src/main/java/com/azure/security/keyvault/administration/KeyVaultBackupAsyncClient.java:343]: nit: simplify to below (same to the one below)?
```suggestion
            .doOnRequest(ignored -> logger.info("Restoring from location - {}", backupFolderUrl))
```

File: sdk/keyvault/azure-security-keyvault-administration/src/main/java/com/azure/security/keyvault/administration/KeyVaultBackupAsyncClient.java [hunk 3]
```diff
@@ -514,23 +517,22 @@ private static LongRunningOperationStatus toLongRunningOperationStatus(String op
...
-                logger.info("Restoring key \"{}\" from location - {}\\{}", keyName, blobStorageUrl, folderName))
+                logger.info("Restoring key \"{}\" from location - {}{}", keyName, containerUrl, folderName))
             .doOnSuccess(response ->
-                logger.info("Restored key \"{}\" from location - {}\\{}", keyName, blobStorageUrl, folderName))
+                logger.info("Restored key \"{}\" from location - {}{}", keyName, containerUrl, folderName))
             .doOnError(error ->
-                logger.warning("Failed to restore key \"{}\" from location - {}\\{}", keyName, blobStorageUrl,
-                    folderName, error))
+                logger.warning("Failed to restore key \"{}\" from location - {}{}", keyName, containerUrl, folderName,
+                    error))
```
Comment (christothes) [inline sdk/keyvault/azure-security-keyvault-administration/src/main/java/com/azure/security/keyvault/administration/KeyVaultBackupAsyncClient.java:525]: same comments as above.

File: sdk/keyvault/azure-security-keyvault-administration/src/main/java/com/azure/security/keyvault/administration/KeyVaultBackupAsyncClient.java [hunk 4]
```diff
@@ -599,7 +601,7 @@ private static LongRunningOperationStatus toLongRunningOperationStatus(String op
                 fullBackupOperation.getStartTime(), fullBackupOperation.getEndTime(),
                 fullBackupOperation.getAzureStorageBlobContainerUri());
         } else {
-            throw new IllegalArgumentException("Operation type not supported");
+            throw new UnsupportedOperationException("Operation type not supported");
```
Comment (christothes) [inline sdk/keyvault/azure-security-keyvault-administration/src/main/java/com/azure/security/keyvault/administration/KeyVaultBackupAsyncClient.java:604]: nit: Is it possible to show the name of the type here? Now that the exception type is `UnsupportedOperationException`, the string doesn't add any additional information.

**Review by christothes — APPROVED** [review]

**Pull request merged**